File: backend/internal/features/audit/domain/audit.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

const (
	EventLoginSuccess     = "auth.login.success"
	EventLoginFailure     = "auth.login.failure"
	EventLoginLocked      = "auth.login.locked"
	EventTokenRefresh     = "auth.token.refresh"
	EventLogout           = "auth.logout"
	EventBadgeReissued    = "auth.badge.reissued"
	EventPINChanged       = "auth.pin.changed"
	EventUserCreated      = "user.created"
	EventUserUpdated      = "user.updated"
	EventUserDeleted      = "user.deleted"
	EventUserUnlocked     = "user.unlocked"
	EventPatientCreated  = "patient.created"
	EventPatientUpdated  = "patient.updated"
	EventPatientDeleted  = "patient.deleted"
	EventPatientViewed   = "patient.viewed"
	EventPatientNoteCreated = "patient.note.created"
	EventPatientNoteUpdated = "patient.note.updated"
	EventPatientNoteDeleted = "patient.note.deleted"
	EventPatientDocumentUploaded = "patient.document.uploaded"
	EventPatientDocumentViewed   = "patient.document.viewed"
	EventPatientDocumentDeleted  = "patient.document.deleted"
	EventPatientExercisePlanCreated  = "patient.exercise_plan.created"
	EventPatientExercisePlanUpdated  = "patient.exercise_plan.updated"
	EventPatientExercisePlanDeleted  = "patient.exercise_plan.deleted"
	EventPatientExerciseEntryCreated = "patient.exercise_entry.created"
	EventPatientExerciseEntryUpdated = "patient.exercise_entry.updated"
	EventPatientExerciseEntryDeleted = "patient.exercise_entry.deleted"
	EventFormSubmissionCreated   = "form.submission.created"
	EventBackupCreated    = "system.backup.created"
	EventBackupRestored   = "system.backup.restored"
	EventHealthCheckRun   = "system.healthcheck.run"
	EventComplianceCheckRun = "system.compliance.run"
	EventAuditPurged      = "system.audit.purged"
)

// Event represents a HIPAA-relevant audit log entry.
type Event struct {
	ID           uuid.UUID
	EventType    string
	UserID       *uuid.UUID
	ResourceType string
	ResourceID   string
	IPAddress    string
	UserAgent    string
	Details      map[string]string
	Success      bool
	CreatedAt    time.Time
}

// Repository persists audit events.
type Repository interface {
	Create(ctx context.Context, event *Event) error
	ListForUser(ctx context.Context, userID string, page, pageSize int) ([]*Event, int64, error)
	ListRecent(ctx context.Context, filter ListFilter) ([]*Event, int64, error)
}

// ListFilter narrows a system-wide audit query.
type ListFilter struct {
	Page      int
	PageSize  int
	EventType string
	UserID    string
	SinceDays int
}

// Logger records audit events.
type Logger interface {
	Log(ctx context.Context, event Event)
}

File: frontend/src/types/audit.ts
export interface AuditEntry {
  id: string;
  eventType: string;
  userId?: string;
  resourceType?: string;
  resourceId?: string;
  ipAddress?: string;
  userAgent?: string;
  details?: Record<string, string>;
  success: boolean;
  createdAt: string;
}

export interface PaginatedAuditEntries {
  events: AuditEntry[];
  totalCount: number;
  currentPage: number;
  pageSize: number;
  totalPages: number;
}

export const AUDIT_EVENT_LABELS: Record<string, string> = {
  'auth.login.success': 'Signed in',
  'auth.login.failure': 'Failed sign in',
  'auth.login.locked': 'Account locked',
  'auth.token.refresh': 'Session refreshed',
  'auth.logout': 'Signed out',
  'auth.badge.reissued': 'Badge reissued',
  'auth.pin.changed': 'PIN changed',
  'user.created': 'Account created',
  'user.updated': 'Profile updated',
  'user.deleted': 'Account deleted',
  'user.unlocked': 'Account unlocked',
  'patient.created': 'Patient created',
  'patient.updated': 'Patient updated',
  'patient.deleted': 'Patient deleted',
  'patient.viewed': 'Patient viewed',
  'patient.note.created': 'Clinical note added',
  'patient.note.updated': 'Clinical note edited',
  'patient.note.deleted': 'Clinical note deleted',
  'patient.document.uploaded': 'Document uploaded',
  'patient.document.viewed': 'Document downloaded',
  'patient.document.deleted': 'Document removed',
  'system.backup.created': 'Backup created',
  'system.backup.restored': 'Database restored',
  'system.healthcheck.run': 'Health check run',
  'system.compliance.run': 'Compliance check run',
  'system.audit.purged': 'Audit log rotation',
};

export function formatAuditEvent(entry: AuditEntry): string {
  const base = AUDIT_EVENT_LABELS[entry.eventType] ?? entry.eventType;
  const details = entry.details ?? {};

  if (entry.eventType === 'auth.login.success' && details.loginMode) {
    return `${base} (${details.loginMode})`;
  }
  if (entry.eventType === 'auth.login.failure' && details.reason) {
    return `${base}: ${details.reason.replace(/_/g, ' ')}`;
  }
  if (entry.eventType === 'auth.badge.reissued' && details.actorId) {
    return `${base} by administrator`;
  }
  if (entry.eventType.startsWith('patient.') && entry.resourceId) {
    return `${base}`;
  }

  return base;
}
